[TASK] Add rector to composer.json with a simple null-config

ziegenberg opened the pull request:
Adding rector as a development dependency enables us to do automatic refactorings. Initially, this comes with a simple null-config, that does not do much but run rector, so following the suggested integration path of "taking it slow" [1].

Helps with #432

[1] https://getrector.com/documentation/integration-to-new-project

oliverklee commented on composer.json:
```diff
@@ -31,7 +31,8 @@
         "phpstan/phpstan-phpunit": "^1.4.0",
-        "phpunit/phpunit": "^8.5.38"
+        "phpunit/phpunit": "^8.5.38",
+        "rector/rector": "^1.1"
```
For development dependencies, please let's always use a specific version number:

```suggestion
        "rector/rector": "^1.1.0"
```

ziegenberg replied:
Fixed it to the suggestesd v1.1.0

oliverklee commented on composer.json:
```diff
@@ -70,32 +71,38 @@
         ],
         "ci:php:fixer": "\"./.phive/php-cs-fixer\" --config=config/php-cs-fixer.php fix --dry-run -v --show-progress=dots --diff bin src tests",
         "ci:php:stan": "phpstan --no-progress --configuration=config/phpstan.neon",
+        "ci:php:rector": "\"./vendor/bin/rector\" --no-progress-bar --no-diffs --dry-run --config=config/rector.php",
```
```suggestion
        "ci:php:rector": "rector --no-progress-bar --no-diffs --dry-run --config=config/rector.php",
```

Also, I think that having the diff might be helpful to the human reader. What do you think?

ziegenberg replied:
Fixed.

oliverklee commented on composer.json:
```diff
@@ -70,32 +71,38 @@
             "@ci:php:fixer",
-            "@ci:php:stan"
+            "@ci:php:stan",
+            "@ci:php:rector"
```
Please let's sort those lines alphabetically (says my inner Monk ;-) ).

ziegenberg replied:
Sorted!

oliverklee commented on composer.json:
```diff
@@ -70,32 +71,38 @@
         "fix:php": [
-            "@fix:php:fixer"
+            "@fix:php:fixer",
+            "@ci"
```
This probably should be `"@fix:php:rector"`, I'd guess?

ziegenberg replied:
Yes, you're right.

oliverklee reviewed (changes requested):
Thanks for contributing this PR, @ziegenberg! :heart:

ziegenberg commented:
I addressed your comments, rebased onto the current main and force-pushed my changes. Please let me know if you would rather have fixup commits for later squashing them.

oliverklee commented on composer.json:
```diff
@@ -70,8 +71,10 @@
         ],
         "ci:php:fixer": "\"./.phive/php-cs-fixer\" --config=config/php-cs-fixer.php fix --dry-run -v --show-progress=dots --diff bin src tests",
         "ci:php:stan": "phpstan --no-progress --configuration=config/phpstan.neon",
+        "ci:php:rector": "\"./vendor/bin/rector\" --no-progress-bar --dry-run --config=config/rector.php",
```
For binaries supplied by Composer packages, we can use them directly in Composer commands without the `vendor/bin` path:

```suggestion
        "ci:php:rector": "rector --no-progress-bar --dry-run --config=config/rector.php",
```

oliverklee commented:
> Please let me know if you would rather have fixup commits for later squashing them.

Yes, I'd prefer that as that makes re-reviewing a PR a bit easier. Thanks for asking!

ziegenberg commented:
I addressed the last two comments regarding the Composer commands in a fixup commit. If you're done reviewing I can quash them.

oliverklee commented:
> If you're done reviewing I can quash them.

Thanks! There's no need, though - GitHub allows me to squash all commits of a PR when I merge it.

oliverklee commented on composer.json:
```diff
@@ -80,22 +83,26 @@
+            "@fix:php:rector"
         ],
         "fix:php:fixer": "\"./.phive/php-cs-fixer\" --config=config/php-cs-fixer.php fix bin src tests",
+        "fix:php:rector": "vendor/bin/rector --config=config/rector.php",
```
```suggestion
        "fix:php:rector": "rector --config=config/rector.php",
```

ziegenberg replied:
And also fixed.

oliverklee reviewed (approved):
Looks good. Thanks! :pray:
